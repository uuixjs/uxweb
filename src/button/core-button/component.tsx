import { CoreInteractive, CoreInteractiveElement, CoreInteractivePublicProps } from "../../core-interactive";
import { ForwardRefRenderFunction, ReactNode, forwardRef } from "react";
import {
  Interpolation,
  ThemeContextValue,
  ThemeProps,
  css,
  focusVisible,
  hoverCss,
  staticTokenRule,
  styled,
  themeTokenRule,
} from "@uuixjs/uuixweb-lib";

import { StaticToken } from "@uuixjs/uuixweb-lib";
import { withOverlayContext } from "../../overlay-region/context";

export enum CoreButtonSize {
  Default = "default",
  Small = "small",
  Large = "large",
}

export enum CoreButtonType {
  Primary = "primary",
  Secondary = "secondary",
  Text = "text",
  Destructive = "destructive",
  Success = "success",
}

export enum CoreButtonRounding {
  Default = "default",
  Left = "left",
  Right = "right",
}

type CoreButtonRoundingValue = "default" | "left" | "right"; // TODO: https://jira.xarth.tv/browse/COREUI-3554

export interface CoreButtonPublicProps extends CoreInteractivePublicProps {
  fullWidth?: boolean;
  size?: CoreButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

export interface CoreButtonProps extends CoreButtonPublicProps {
  children?: ReactNode;
  className?: string;
  rounding?: CoreButtonRounding | CoreButtonRoundingValue; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  overlay?: boolean;
  variant?: CoreButtonType | "primary" | "secondary" | "text" | "destructive" | "success"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

const getRoundingStyles = (rounding: CoreButtonRoundingValue, borderRadius: StaticToken) => {
  if (rounding === CoreButtonRounding.Left) {
    return css`
      border-top-left-radius: ${staticTokenRule(borderRadius)};
      border-bottom-left-radius: ${staticTokenRule(borderRadius)};
    `;
  }

  if (rounding === CoreButtonRounding.Right) {
    return css`
      border-top-right-radius: ${staticTokenRule(borderRadius)};
      border-bottom-right-radius: ${staticTokenRule(borderRadius)};
    `;
  }

  return css`
    border-radius: ${staticTokenRule(borderRadius)};
  `;
};

const getSizeStyles = ({
  $size = CoreButtonSize.Default,
  $rounding = CoreButtonRounding.Default,
}: ScCoreButtonProps) => {
  let borderRadius: StaticToken = "border-radius-medium";
  let fontSize: StaticToken = "button-text-default";
  let height: StaticToken = "button-size-default";

  if ($size === CoreButtonSize.Small) {
    borderRadius = "border-radius-small";
    fontSize = "button-text-small";
    height = "button-size-small";
  }

  if ($size === CoreButtonSize.Large) {
    borderRadius = "border-radius-large";
    fontSize = "button-text-large";
    height = "button-size-large";
  }

  return css`
    ${getRoundingStyles($rounding, borderRadius)}
    font-size: ${staticTokenRule(fontSize)};
    height: ${staticTokenRule(height)};
  `;
};

type StyleValues<P> = Interpolation<P & ThemeProps<ThemeContextValue>>;

const baseStyles = (styles: StyleValues<ThemeProps<ThemeContextValue>>) => (props: ScCoreButtonProps) =>
  !props.$disabled && styles;

const hoverStyles = (styles: StyleValues<ThemeProps<ThemeContextValue>>) => (props: ScCoreButtonProps) =>
  !props.$disabled && hoverCss`${styles}`;

const activeStyles = (styles: StyleValues<ThemeProps<ThemeContextValue>>) => (props: ScCoreButtonProps) =>
  !props.$disabled &&
  css`
    &:active {
      ${styles}
    }
    ${hoverCss`
      &:active {
        ${styles}
      }
    `}
  `;

const disabledStyles = (styles: StyleValues<ThemeProps<ThemeContextValue>>) => (props: ScCoreButtonProps) =>
  props.$disabled && styles;

export interface ScCoreButtonProps {
  $disabled?: boolean;
  $fullWidth?: boolean;
  $overlay?: boolean;
  $rounding?: CoreButtonRounding | CoreButtonRoundingValue; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  $size?: CoreButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

export const ScCoreButton = styled(CoreInteractive)<ScCoreButtonProps>`
  display: inline-flex;
  position: relative;
  align-items: center;
  justify-content: center;
  vertical-align: middle;
  overflow: hidden;
  text-decoration: none;
  white-space: nowrap;
  user-select: none;
  width: ${(props) => (props.$fullWidth ? "100%" : undefined)};
  font-weight: ${staticTokenRule("font-weight-semibold")};

  ${(props) => getSizeStyles(props)}

  ${(props) =>
    focusVisible`
      box-shadow: ${themeTokenRule(props.$overlay ? "shadow-button-overlay-focus" : "shadow-button-focus")};
    `}

  /* Base styles reset */
  &:hover {
    text-decoration: none;
  }

  /* Includes disableInteraction */
  &:disabled {
    cursor: not-allowed;
  }

  opacity: ${(props) => props.$disabled && 0.5};
`;

const ScCoreButtonPrimary = styled(ScCoreButton)<ScCoreButtonProps>((props) => {
  if (props.$overlay) {
    return css`
      background-color: ${themeTokenRule("color-background-button-overlay-primary-default")};
      color: ${themeTokenRule("color-text-button-overlay-primary")};
      ${hoverStyles(css`
        background-color: ${themeTokenRule("color-background-button-overlay-primary-hover")};
        color: ${themeTokenRule("color-text-button-overlay-primary")};
      `)}
      ${focusVisible`
        background-color: ${themeTokenRule("color-background-button-overlay-primary-hover")};
      `}
      ${activeStyles(css`
        background-color: ${themeTokenRule("color-background-button-overlay-primary-active")};
        box-shadow: ${themeTokenRule("shadow-button-overlay-active")};
      `)}
    `;
  } else {
    return css`
      background-color: ${themeTokenRule("color-background-button-primary-default")};
      color: ${themeTokenRule("color-text-button-primary")};
      ${hoverStyles(css`
        background-color: ${themeTokenRule("color-background-button-primary-hover")};
        color: ${themeTokenRule("color-text-button-primary")};
      `)}
      ${focusVisible`
        background-color: ${themeTokenRule("color-background-button-primary-hover")};
      `}
      ${activeStyles(css`
        background-color: ${themeTokenRule("color-background-button-primary-active")};
      `)}
      ${disabledStyles(css`
        background-color: ${themeTokenRule("color-background-button-disabled")};
        color: ${themeTokenRule("color-text-button-disabled")};
      `)}
    `;
  }
});

const ScCoreButtonSecondary = styled(ScCoreButton)<ScCoreButtonProps>((props) => {
  if (props.$overlay) {
    return css`
      background-color: ${themeTokenRule("color-background-button-overlay-secondary-default")};
      color: ${themeTokenRule("color-text-button-overlay-secondary")};
      ${hoverStyles(css`
        background-color: ${themeTokenRule("color-background-button-overlay-secondary-hover")};
        color: ${themeTokenRule("color-text-button-overlay-secondary")};
      `)}
      ${focusVisible`
        background-color: ${themeTokenRule("color-background-button-overlay-secondary-hover")};
      `}
      ${activeStyles(css`
        background-color: ${themeTokenRule("color-background-button-overlay-secondary-active")};
        box-shadow: ${themeTokenRule("shadow-button-overlay-active")};
      `)}
    `;
  } else {
    return css`
        background-color: ${themeTokenRule("color-background-button-secondary-default")};
        color: ${themeTokenRule("color-text-button-secondary")};
        ${hoverStyles(css`
          background-color: ${themeTokenRule("color-background-button-secondary-hover")};
          color: ${themeTokenRule("color-text-button-secondary")};
        `)}
        ${focusVisible`
          background-color: ${themeTokenRule("color-background-button-secondary-hover")};
        `}
        ${activeStyles(css`
          background-color: ${themeTokenRule("color-background-button-secondary-active")};
        `)}
        ${disabledStyles(css`
          background-color: ${themeTokenRule("color-background-button-disabled")};
          color: ${themeTokenRule("color-text-button-disabled")};
        `)}
      `;
  }
});

const ScCoreButtonText = styled(ScCoreButton)<ScCoreButtonProps>((props) => {
  if (props.$overlay) {
    return css`
      ${baseStyles(css`
        background-color: ${themeTokenRule("color-background-button-overlay-text-default")};
        color: ${themeTokenRule("color-text-button-overlay-text")};
      `)}
      ${hoverStyles(css`
        background-color: ${themeTokenRule("color-background-button-overlay-text-hover")};
        color: ${themeTokenRule("color-text-button-overlay-text")};
      `)}
      ${focusVisible`
        background-color: ${themeTokenRule("color-background-button-overlay-text-hover")};
      `}
      ${activeStyles(css`
        background-color: ${themeTokenRule("color-background-button-overlay-text-active")};
        box-shadow: ${themeTokenRule("shadow-button-overlay-active")};
      `)}
    `;
  } else {
    return css`
      ${baseStyles(css`
        background-color: ${themeTokenRule("color-background-button-text-default")};
        color: ${themeTokenRule("color-text-button-text")};
      `)}
      ${hoverStyles(css`
        background-color: ${themeTokenRule("color-background-button-text-hover")};
      `)}
      ${focusVisible`
        background-color: ${themeTokenRule("color-background-button-text-hover")};
      `}
      ${activeStyles(css`
        background-color: ${themeTokenRule("color-background-button-text-active")};
      `)}
      ${disabledStyles(css`
        color: ${themeTokenRule("color-text-button-disabled")};
      `)}
    `;
  }
});

const ScCoreButtonDestructive = styled(ScCoreButton)<ScCoreButtonProps>`
  ${baseStyles(css`
    background-color: ${themeTokenRule("color-background-button-destructive-default")};
    color: ${themeTokenRule("color-text-button-destructive")};
  `)}
  ${hoverStyles(css`
    background-color: ${themeTokenRule("color-background-button-destructive-hover")};
    color: ${themeTokenRule("color-text-button-destructive")};
  `)}
  ${focusVisible`
    background-color: ${themeTokenRule("color-background-button-destructive-hover")};
  `}
  ${activeStyles(css`
    background-color: ${themeTokenRule("color-background-button-destructive-active")};
  `)}
  ${(props) =>
    disabledStyles(
      props.$overlay
        ? css`
            background-color: ${themeTokenRule("color-background-button-overlay-primary-default")};
            color: ${themeTokenRule("color-text-button-overlay-primary")};
          `
        : css`
            background-color: ${themeTokenRule("color-background-button-disabled")};
            color: ${themeTokenRule("color-text-button-disabled")};
          `,
    )}
`;

export const ScCoreButtonSuccess = styled(ScCoreButton)<ScCoreButtonProps>`
    ${baseStyles(css`
      background-color: ${themeTokenRule("color-background-button-success")};
      color: ${themeTokenRule("color-text-button-success")};
    `)}
    ${hoverStyles(css`
      background-color: ${themeTokenRule("color-background-button-success-hover")};
      color: ${themeTokenRule("color-text-button-success")};
    `)}
    ${focusVisible`
      background-color: ${themeTokenRule("color-background-button-success-hover")}
    `}
    ${activeStyles(
      css`
        background-color: ${themeTokenRule("color-background-button-success-active")};
      `,
    )}
    ${(props) =>
      disabledStyles(
        props.$overlay
          ? css`
              background-color: ${themeTokenRule("color-background-button-overlay-primary-default")};
              color: ${themeTokenRule("color-text-button-overlay-primary")};
            `
          : css`
              background-color: ${themeTokenRule("color-background-button-disabled")};
              color: ${themeTokenRule("color-text-button-disabled")};
            `,
      )}
`;

export const CoreButtonComponent: ForwardRefRenderFunction<CoreInteractiveElement, CoreButtonProps> = (
  {
    children,
    className,
    disabled,
    fullWidth,
    overlay,
    rounding = CoreButtonRounding.Default,
    size = CoreButtonSize.Default,
    variant,
    ...props
  },
  ref,
) => {
  const componentProps = {
    $disabled: disabled,
    $fullWidth: fullWidth,
    $overlay: overlay,
    $rounding: rounding,
    $size: size,
    children,
    className,
    disabled,
    ref,
    ...props,
  };

  if (variant === CoreButtonType.Primary) {
    return <ScCoreButtonPrimary {...componentProps} />;
  }

  if (variant === CoreButtonType.Secondary) {
    return <ScCoreButtonSecondary {...componentProps} />;
  }

  if (variant === CoreButtonType.Text) {
    return <ScCoreButtonText {...componentProps} />;
  }

  if (variant === CoreButtonType.Destructive) {
    return <ScCoreButtonDestructive {...componentProps} />;
  }

  if (variant === CoreButtonType.Success) {
    return <ScCoreButtonSuccess {...componentProps} />;
  }

  return <ScCoreButton {...componentProps} />;
};

CoreButtonComponent.displayName = "CoreButton";
const ComponentWithRef = forwardRef(CoreButtonComponent);
const ComponentWithOverlayContext = withOverlayContext<CoreButtonProps, HTMLElement>(ComponentWithRef);
export { ComponentWithOverlayContext as CoreButton };
