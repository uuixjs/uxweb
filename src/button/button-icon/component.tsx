import { ButtonHTMLAttributes, ForwardRefRenderFunction, forwardRef } from "react";
import { CoreButton, CoreButtonSize, CoreButtonType } from "../core-button";
import { CoreInteractiveElement, CoreInteractivePublicProps } from "../../core-interactive";
import {
  css,
  focusVisible,
  hoverCss,
  rem,
  staticTokenRule,
  styleVariant,
  styled,
  themeTokenRule,
} from "@uuixjs/uuixweb-lib";

import { Icon } from "../../icon";
import { SVGAsset } from "../../svg";
import { StatusAlertButton } from "../status-alert-button";
import { withOverlayContext } from "../../overlay-region/context";

export enum ButtonIconType {
  Primary = "primary",
  // @deprecated Visually identical to primary; please us primary instead.
  Secondary = "secondary",
  Alert = "alert",
  Success = "success",
}

export enum ButtonIconSize {
  Default = "default",
  Small = "small",
  Large = "large",
}

const BUTTON_TYPES: { [key: string]: CoreButtonType } = {
  [ButtonIconType.Alert]: CoreButtonType.Destructive,
  [ButtonIconType.Success]: CoreButtonType.Success,
};

const BUTTON_SIZES: { [key: string]: CoreButtonSize } = {
  [ButtonIconSize.Small]: CoreButtonSize.Small,
  [ButtonIconSize.Default]: CoreButtonSize.Default,
  [ButtonIconSize.Large]: CoreButtonSize.Large,
};

const BUTTON_ICON_SIZES = {
  [ButtonIconSize.Small]: { width: 16, height: 16 },
  [ButtonIconSize.Default]: { width: 20, height: 20 },
  [ButtonIconSize.Large]: { width: 24, height: 24 },
};

export const REVERSE_BUTTON_SIZES: { [key: string]: ButtonIconSize } = {
  [CoreButtonSize.Small]: ButtonIconSize.Small,
  [CoreButtonSize.Default]: ButtonIconSize.Default,
  [CoreButtonSize.Large]: ButtonIconSize.Large,
};

export interface ButtonIconProps
  extends CoreInteractivePublicProps,
    // Adding ButtonHTMLAttributes for "type" is a workaround for now to address a common use case; this component may actually render an <a> tag or a <button>!
    Pick<ButtonHTMLAttributes<HTMLButtonElement>, "type"> {
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refHandler` */
  refHandler?: CoreInteractivePublicProps["refHandler"];
  "aria-label": string;
  dropdown?: boolean;
  overlay?: boolean;
  icon: SVGAsset | JSX.Element;
  size?: ButtonIconSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  statusAlertIcon?: SVGAsset;
  variant?: ButtonIconType | "primary" | "secondary" | "alert" | "success"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

export interface ScButtonIconProps {
  $dropdown?: ButtonIconProps["dropdown"];
}

interface ScButtonIconFigureProps {
  size?: ButtonIconProps["size"];
  icon: ButtonIconProps["icon"];
}
export const ScButtonIcon = styled(CoreButton).attrs((props: ScButtonIconFigureProps) => ({
  size: props.size || ButtonIconSize.Default,
}))<ScButtonIconProps>`
  display: inline-flex;
  align-items: center;
  justify-content: center;
  user-select: none;

  height: ${styleVariant("size", {
    [CoreButtonSize.Small]: staticTokenRule("button-size-small"),
    [CoreButtonSize.Default]: staticTokenRule("button-size-default"),
    [CoreButtonSize.Large]: staticTokenRule("button-size-large"),
  })};

  width: ${({ $dropdown }) =>
    !$dropdown &&
    styleVariant("size", {
      [CoreButtonSize.Small]: staticTokenRule("button-size-small"),
      [CoreButtonSize.Default]: staticTokenRule("button-size-default"),
      [CoreButtonSize.Large]: staticTokenRule("button-size-large"),
    })};

  padding: ${({ $dropdown }) =>
    $dropdown &&
    styleVariant("size", {
      [CoreButtonSize.Small]: "0.4rem",
      [CoreButtonSize.Default]: "0.5rem",
      [CoreButtonSize.Large]: "0.6rem",
    })};

  ${({ disabled, variant, overlay }) => {
    if (disabled) {
      return `
        opacity: 0.5;
        cursor: not-allowed;
      `;
    }
    if (overlay && !variant) {
      return css`
        border: ${staticTokenRule("border-width-default")} solid transparent;
        background-color: ${themeTokenRule("color-background-button-icon-overlay-default")};
        color: ${themeTokenRule("color-text-button-overlay")};

        ${hoverCss`
          background-color: ${themeTokenRule("color-background-button-icon-overlay-hover")};
          color: ${themeTokenRule("color-text-button-overlay-hover")};
        `}

        ${focusVisible`
          background-color: ${themeTokenRule("color-background-button-icon-overlay-hover")};
          color: ${themeTokenRule("color-text-button-overlay-focus")};
          box-shadow: ${themeTokenRule("shadow-button-overlay-focus")};
        `}

        &:active {
          background-color: var(--color-background-button-icon-overlay-active);
          color: var(--color-text-button-overlay-active);
          box-shadow: var(--shadow-button-overlay-active);
        }
      `;
    }
    if (!variant) {
      return css`
        border-radius: ${staticTokenRule("border-radius-medium")};
        background-color: ${themeTokenRule("color-background-button-text-default")};
        color: ${themeTokenRule("color-fill-button-icon")};

        ${hoverCss`
          background-color: ${themeTokenRule("color-background-button-text-hover")};
          color: ${themeTokenRule("color-fill-button-icon-hover")};
        `}

        ${focusVisible`
          background-color: ${themeTokenRule("color-background-button-text-hover")};
          color: ${themeTokenRule("color-fill-button-icon-hover")};
          box-shadow: ${themeTokenRule("shadow-button-focus")};
        `}

        &:active {
          background-color: ${themeTokenRule("color-background-button-text-active")};
          color: ${themeTokenRule("color-fill-button-icon-active")};
        }
      `;
    }
  }}
`;
export const ScButtonIconFigure = styled.div.attrs((props: ScButtonIconFigureProps) => ({
  size: props.size || ButtonIconSize.Default,
  children: typeof props.icon === "string" ? <Icon fill asset={props.icon} /> : props.icon,
}))<ScButtonIconFigureProps>`
  ${({ size }) => ({
    width: rem(BUTTON_ICON_SIZES[size || ButtonIconSize.Default].width),
    height: rem(BUTTON_ICON_SIZES[size || ButtonIconSize.Default].height),
  })}
`;

export const ButtonIconComponent: ForwardRefRenderFunction<CoreInteractiveElement, ButtonIconProps> = (
  { disabled, variant, statusAlertIcon, size, icon, overlay, dropdown, ...props },
  ref,
) => {
  const buttonIconFigureProps = {
    size,
    icon,
  };

  if (statusAlertIcon && typeof icon === "string") {
    return <StatusAlertButton {...props} disabled={disabled} ref={ref} icon={icon} statusAlertIcon={statusAlertIcon} />;
  }

  return (
    <ScButtonIcon
      {...props}
      ref={ref}
      variant={variant && !overlay ? BUTTON_TYPES[variant] : undefined}
      disabled={disabled}
      overlay={overlay}
      size={size ? BUTTON_SIZES[size] : CoreButtonSize.Default}
      $dropdown={dropdown}
    >
      <ScButtonIconFigure {...buttonIconFigureProps} />
      {dropdown && <ScButtonIconFigure size={size} icon={SVGAsset.GlyphArrDown} />}
    </ScButtonIcon>
  );
};

ButtonIconComponent.displayName = "ButtonIcon";
const ComponentWithRef = forwardRef(ButtonIconComponent);
const ComponentWithOverlayContext = withOverlayContext<ButtonIconProps, HTMLElement>(ComponentWithRef);
export { ComponentWithOverlayContext as ButtonIcon };
