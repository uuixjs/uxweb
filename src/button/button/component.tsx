import {
  ButtonHTMLAttributes,
  ForwardRefRenderFunction,
  ReactNode,
  forwardRef,
} from "react";
import {
  CoreButton,
  CoreButtonSize,
  CoreButtonType,
} from "../core-button";
import {
  CoreInteractiveElement,
  CoreInteractivePublicProps,
} from "../../core-interactive";
import { LoadingButton, LoadingStatus } from "../loading-button";
import { PurchaseButton, PurchaseButtonProps } from "../purchase-button";

import { CoreButtonLabel } from "../core-button/core-button-label";
import { SVGAsset } from "../../svg";
import { StatusAlertButton } from "../status-alert-button";

export enum ButtonType {
  Primary = "primary",
  Secondary = "secondary",
  Text = "text",
  Alert = "alert",
  Success = "success",
  Warning = "warning",
  Info = "info",
}

export enum ButtonSize {
  Default = "default",
  Small = "small",
  Large = "large",
}

export enum ButtonState {
  Default = "default",
  Loading = "loading",
  Success = "success",
}

export interface ButtonPublicProps extends CoreInteractivePublicProps {
  /** @deprecated Ref forwarding is now supported; use the `ref` prop instead of `refHandler` */
  refHandler?: CoreInteractivePublicProps["refHandler"];
  /**
   * The content rendered within the button. This includes the button text.
   *
   * @example Chat
   */
  children?: ReactNode;
  /** Indicates the button will open a dropdown. */
  dropdown?: boolean;
  /** Sets the button to stretch across the width of its container. */
  fullWidth?: boolean;
  /** Displays an icon in the button. */
  icon?: SVGAsset;
  /** Indicates that this button overlays on a darker background. */
  overlay?: boolean;
  /** Size of the button. */
  size?: ButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554

  /** The type of button to display. */
  variant?: ButtonType | "primary" | "secondary" | "text" | "alert" | "success"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

export interface ButtonStatusAlertButtonProps
  extends CoreInteractivePublicProps {
  /** Icon to display when the status button is hovered.
   *
   * @deprecated use `<StatusAlertButton>` instead.
   */
  statusAlertIcon?: SVGAsset;
  /**
   * Text to display when the status button is hovered. Setting this property
   * turns the button into a "status" button.
   *
   * @deprecated use `<StatusAlertButton>` instead.
   */
  statusAlertText?: string;
  /**
   * The content rendered within the button. This includes the button text.
   *
   * @example Chat
   */
  children: ReactNode;
  /** Size of the button. */
  size?: ButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554

  /** The type of button to display. */
  variant?: ButtonType | "primary" | "secondary" | "text" | "alert" | "success"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /** Displays an icon in the button. */
  icon?: SVGAsset;
  /** Sets the button to stretch across the width of its container. */
  fullWidth?: boolean;
}

export interface ButtonPurchaseButtonProps extends CoreInteractivePublicProps {
  /** Displays a price associated with the button action.
   *
   * @deprecated use PurchaseButton instead.
   */
  purchase?: string;
  /**
   * The content rendered within the button. This includes the button text.
   *
   * @example Chat
   */
  children: ReactNode;
  /** Size of the button. */
  size?: ButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /** Displays an icon in the button. */
  icon?: SVGAsset;
  /** Sets the button to stretch across the width of its container. */
  fullWidth?: boolean;
}

export interface ButtonLoadingButtonProps extends CoreInteractivePublicProps {
  /**
   * @deprecated use `<LoadingButton>` instead.
   */
  state?: ButtonState | "default" | "loading" | "success"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /**
   * The content rendered within the button. This includes the button text.
   *
   * @example Chat
   */
  children: ReactNode;
  /** Size of the button. */
  size?: ButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554

  /** The type of button to display. */
  variant?: ButtonType | "primary" | "secondary" | "text" | "alert" | "success"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
  /** Displays an icon in the button. */
  icon?: SVGAsset;
  /** Sets the button to stretch across the width of its container. */
  fullWidth?: boolean;
}

export type ButtonProps =
  | ButtonPublicProps
  | ButtonStatusAlertButtonProps
  | ButtonPurchaseButtonProps
  | ButtonLoadingButtonProps;

/**
 * Buttons communicate that an action will happen when clicked. Best used as
 * primary actions inside of another component such as a Modals, Cards, and/or
 * forms. **Buttons should never be used for generic links; use CoreLink instead.**
 */
const Button: ForwardRefRenderFunction<
  CoreInteractiveElement,
  ButtonProps &
    // Adding ButtonHTMLAttributes for "type" is a workaround for now to address a common use case; this component may actually render an <a> tag or a <button>!
    Pick<ButtonHTMLAttributes<HTMLButtonElement>, "type">
> = (props, ref) => {
  const BUTTON_TYPES: { [key: string]: CoreButtonType } = {
    [ButtonType.Primary]: CoreButtonType.Primary,
    [ButtonType.Secondary]: CoreButtonType.Secondary,
    [ButtonType.Text]: CoreButtonType.Text,
    [ButtonType.Alert]: CoreButtonType.Destructive,
    [ButtonType.Warning]: CoreButtonType.Warning,
    [ButtonType.Info]: CoreButtonType.Info,
    [ButtonType.Success]: CoreButtonType.Success,
  };

  const BUTTON_SIZES: { [key: string]: CoreButtonSize } = {
    [ButtonSize.Small]: CoreButtonSize.Small,
    [ButtonSize.Default]: CoreButtonSize.Default,
    [ButtonSize.Large]: CoreButtonSize.Large,
  };

  if (isPurchaseProps(props)) {
    return (
      <PurchaseButton
        ref={ref}
        {...props}
        size={BUTTON_SIZES[props.size || ButtonSize.Default]}
        price={props.purchase as PurchaseButtonProps["price"]}
      />
    );
  }

  const CORE_PROPS = {
    size: BUTTON_SIZES[props.size || ButtonSize.Default],
    variant: BUTTON_TYPES[props.variant || ButtonType.Primary],
  };

  if (isStatusAlertButton(props)) {
    return <StatusAlertButton ref={ref} {...props} {...CORE_PROPS} />;
  }

  if (isLoadingButton(props)) {
    const BUTTON_STATE: { [key: string]: LoadingStatus } = {
      [ButtonState.Default]: LoadingStatus.Default,
      [ButtonState.Loading]: LoadingStatus.Loading,
      [ButtonState.Success]: LoadingStatus.Success,
    };

    return (
      <LoadingButton
        ref={ref}
        {...props}
        {...CORE_PROPS}
        loadingStatus={BUTTON_STATE[props.state || LoadingStatus.Default]}
      />
    );
  }

  const {
    dropdown,
    fullWidth,
    icon,
    overlay,
    size,
    variant,
    ...buttonProps
  } = props;

  return (
    <CoreButton
      {...buttonProps}
      fullWidth={fullWidth}
      overlay={overlay}
      ref={ref}
      size={size ? BUTTON_SIZES[size] : undefined}
      variant={variant ? BUTTON_TYPES[variant] : CoreButtonType.Primary}
    >
      <CoreButtonLabel
        dropdown={dropdown}
        icon={icon}
        size={size ? BUTTON_SIZES[size] : undefined}
      >
        {props.children}
      </CoreButtonLabel>
    </CoreButton>
  );
};

Button.displayName = "Button";
const ComponentWithRef = forwardRef(Button);
export { ComponentWithRef as Button };

function isPurchaseProps(
  props: ButtonProps,
): props is ButtonPurchaseButtonProps {
  return (props as ButtonPurchaseButtonProps).purchase !== undefined;
}

function isStatusAlertButton(
  props: ButtonProps,
): props is ButtonStatusAlertButtonProps {
  return (
    ((props as ButtonStatusAlertButtonProps).statusAlertIcon ||
      (props as ButtonStatusAlertButtonProps).statusAlertText) !== undefined
  );
}

function isLoadingButton(
  props: ButtonProps,
): props is ButtonLoadingButtonProps {
  return (props as ButtonLoadingButtonProps).state !== undefined;
}
