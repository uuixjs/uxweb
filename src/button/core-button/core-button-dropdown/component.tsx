import { FC } from "react";
import { SVGAsset } from "../../../svg";
import { CoreButtonSize } from "../component";
import { CoreButtonIcon } from "../core-button-icon";

export enum CoreButtonDropdownType {
  ArrowDown = 1,
  ArrowUp = 2,
  Select = 3,
}

const DropdownType = {
  [CoreButtonDropdownType.ArrowUp]: SVGAsset.AngleUp,
  [CoreButtonDropdownType.ArrowDown]: SVGAsset.AngleDown,
  [CoreButtonDropdownType.Select]: SVGAsset.GlyphArrUpDown,
};

export interface CoreButtonDropdownProps {
  type?: CoreButtonDropdownType;
  size?: CoreButtonSize | "default" | "small" | "large"; // TODO: https://jira.xarth.tv/browse/COREUI-3554
}

/**
 * Buttons communicate that an action will happen when clicked. Best used as
 * primary actions inside of another component such as a Modals, Cards, and/or
 * forms. **Buttons should never be used for generic links.**
 */
export const CoreButtonDropdown: FC<CoreButtonDropdownProps> = (props) => {
  return (
    <CoreButtonIcon
      size={props.size}
      icon={
        props.type
          ? DropdownType[props.type]
          : DropdownType[CoreButtonDropdownType.ArrowDown]
      }
    />
  );
};

CoreButtonIcon.displayName = "CoreButtonDropdown";
