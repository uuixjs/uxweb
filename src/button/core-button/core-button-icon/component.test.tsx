import { setupShallowTest } from "../../../../tests/helpers";
import { SVGAsset } from "../../../svg";
import { CoreButtonSize } from "../component";
import { CoreButtonIcon, CoreButtonIconProps } from "./component";

function getOptionalProps(): CoreButtonIconProps {
  return {
    icon: SVGAsset.Gear,
    size: CoreButtonSize.Large,
  };
}

const setupShallow = setupShallowTest(CoreButtonIcon);
describe("CoreButtonIcon", () => {
  it("renders CoreButtonIcon component with required props", () => {
    const { wrapper } = setupShallow();
    expect(wrapper).toMatchSnapshot();
  });

  it("renders CoreButtonIcon component with all props", () => {
    const { wrapper } = setupShallow(getOptionalProps());
    expect(wrapper).toMatchSnapshot();
  });
});
