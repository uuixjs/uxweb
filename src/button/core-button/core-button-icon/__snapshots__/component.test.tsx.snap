// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreButtonIcon renders CoreButtonIcon component with all props 1`] = `
<styled.div
  className="tw-core-button-icon"
  size="2.4rem"
>
  <Icon
    asset="Gear"
    data-a-selector="tw-core-button-icon"
    fill={true}
  />
</styled.div>
`;

exports[`CoreButtonIcon renders CoreButtonIcon component with required props 1`] = `
<styled.div
  className="tw-core-button-icon"
  size="2rem"
>
  <Icon
    data-a-selector="tw-core-button-icon"
    fill={true}
  />
</styled.div>
`;
